tests/unit/test_dns_client.cpp

```diff
@@ -1723,13 +1811,204 @@ void TestDnssdServerProxyCallback(void)
+
+    //- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+    // Start SRP client.
+
+    srpClient->EnableAutoStartMode(nullptr, nullptr);
+    VerifyOrQuit(srpClient->IsAutoStartModeEnabled());
+
+    AdvanceTime(2000);
+    VerifyOrQuit(srpClient->IsRunning());
+
+    writer.Append("otDNS%s.default.service.arpa.", sInstance->Get<Mac::Mac>().GetExtAddress().ToString().AsCString());
+
+    Log("- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - ");
+    Log("SOA Query");
+
+    for (uint8_t iter = 0; iter < 2; iter++)
+    {
+        sQueryRecordInfo.Reset();
+
+        // First iteration: Query for `default.service.arpa.` directly, and
+        // validate that we see the SOA response in the Answer section.
+        // Second iteration: Query for `myhost.default.service.arpa.`, which
+        // should result in no answer but the SOA record in the Additional
+        // section.
+
+        if (iter == 0)
+        {
+            Log("QueryRecord(%s) for SOA RR", "default.service.arpa.");
+            SuccessOrQuit(dnsClient->QueryRecord(Dns::ResourceRecord::kTypeSoa, "default", "service.arpa.",
+                                                 RecordCallback, sInstance));
+        }
+        else
+        {
+            Log("QueryRecord(%s) for SOA RR", "myhost.default.service.arpa.");
+            SuccessOrQuit(dnsClient->QueryRecord(Dns::ResourceRecord::kTypeSoa, "myhost", "default.service.arpa.",
+                                                 RecordCallback, sInstance));
+        }
+
+        AdvanceTime(100);
+        VerifyOrQuit(sQueryRecordInfo.mCallbackCount == 1);
+        SuccessOrQuit(sQueryRecordInfo.mError);
+        VerifyOrQuit(sQueryRecordInfo.mNumRecords == 1);
+
+        if (iter == 0)
+        {
+            VerifyOrQuit(StringMatch(sQueryRecordInfo.mQueryName, "default.service.arpa."));
+            VerifyOrQuit(MapEnum(sQueryRecordInfo.mRecords[0].mSection) == Dns::Client::RecordInfo::kSectionAnswer);
+        }
+        else
+        {
+            VerifyOrQuit(StringMatch(sQueryRecordInfo.mQueryName, "myhost.default.service.arpa."));
+            VerifyOrQuit(MapEnum(sQueryRecordInfo.mRecords[0].mSection) == Dns::Client::RecordInfo::kSectionAuthority);
+        }
+
+        ValidateSoaRecordData(sQueryRecordInfo.mRecords[0], serverName);
+
+        AdvanceTime(1000);
+    }
+
+    Log("- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - ");
+    Log("NS Query");
+
+    for (uint8_t iter = 0; iter < 2; iter++)
+    {
+        sQueryRecordInfo.Reset();
+
+        // First iteration: Query for `default.service.arpa.` directly, and
+        // validate that we see the NS response in the Answer section.
+        // Second iteration: Query for `myhost.default.service.arpa.`, which
+        // should result in no answer but the NS record in the Additional
```

Comment:
Updating based on previous comment, mini-spec point 3:

```suggestion
        // should result in no answer but the SOA record in the Authority
```